compiler/rustc_session/src/config.rs

```diff
@@ -1915,15 +1923,43 @@ fn parse_remap_path_prefix(
     matches: &getopts::Matches,
     error_format: ErrorOutputType,
 ) -> Vec<(PathBuf, PathBuf)> {
-    matches
-        .opt_strs("remap-path-prefix")
+    enum RemapType {
+        PathPrefix(String),
+        CompilationDir(String),
+    }
+    let mut positions = Vec::new();
+    positions.extend(
+        matches
+            .opt_strs_pos("remap-path-prefix")
+            .into_iter()
+            .map(|(i, s)| (i, RemapType::PathPrefix(s))),
+    );
+    positions.extend(
+        matches
+            .opt_strs_pos("debug-compilation-dir")
+            .into_iter()
+            .map(|(i, s)| (i, RemapType::CompilationDir(s))),
+    );
+    positions.sort_by(|a, b| a.0.cmp(&b.0));
+
+    positions
         .into_iter()
-        .map(|remap| match remap.rsplit_once('=') {
-            None => early_error(
-                error_format,
-                "--remap-path-prefix must contain '=' between FROM and TO",
-            ),
-            Some((from, to)) => (PathBuf::from(from), PathBuf::from(to)),
+        .map(|(_, remap_type)| match remap_type {
+            RemapType::PathPrefix(remap) => match remap.rsplit_once('=') {
+                None => early_error(
+                    error_format,
+                    "--remap-path-prefix must contain '=' between FROM and TO",
+                ),
+                Some((from, to)) => (PathBuf::from(from), PathBuf::from(to)),
+            },
+            RemapType::CompilationDir(to) => match std::env::current_dir() {
+                Err(_) => early_error(
+                    error_format,
+                    "Unable to determine current directory for \
+                    --debug-compilation-dir",
+                ),
+                Ok(from) => (PathBuf::from(from), PathBuf::from(to)),
```

Comment:
This only matches the precise cwd, not a trailing slash, so you're required to map to a non-empty path. So `/home/me/src/foo.rs` with a cwd of `/home/me` can only be mapped to `./src/foo.rs`. But if you match against `/home/me/` then you can choose to map it to '' giving `src/foo.rs` which is a bit cleaner. But that means you need to take the current path separator into account.

This implementation also has the pitfall that it will remap `/home/meat/src/bar.rs` to `at/src/bar.rs` since `--remap-path-prefix` works strictly on strings without regard to path separators.

I think there's an opportunity to discuss a more pathname-aware prefix-replacement algorithm, but there are a lot of edge cases to take into account (the current algorithm, for all its faults, is very simple to explain).

Reply:
It’s cleaner without the path separator but there’s no less functionality with it, right?

If you wanted to specify a different prefix you can with something like “/foo/.” which would translate “/home/me/src/file.rs” into “/foo/./src/file.rs”. It’s not how you’d normally write it by hand, but it functions just fine, and is universally deterministic. Is it worth it to you to add extra code in here to avoid some dots in command lines?

> This implementation also has the pitfall that it will remap /home/meat/src/bar.rs to at/src/bar.rs

Yes, true. However as this path will be rooted at the current working directory, all paths have the same prefix, so this shouldn’t actually come up, right?

Is there something we need to change here with the preexisting flag? Or could that be considered separately? The way this new flag is defined, as an alias for —remap-path-prefix, would allow for it to evolve and both command line flags could be changed in tandem.

Reply:
You wouldn't be able to map a path to `src/foo.rs` with no prefix, and random intermixing of ``, `./` or even `././` gets messy esp if you're trying to be careful because of exact string matching. But no, not the end of the world.